samples: flash_shell: new commands write_unaligned and write_pattern

agansari opened the pull request:
Working with large flash areas (256 or 512 bytes) that write 1 page at a time is hard using flash_shell.

This pull request adds 3 new functionalities:
 - write_unaligned - automated writing to any address, backs up flash to buffers, overwrites buffers, erases affected pages, writes backup buffers to erased pages. It's automated flash writing, user does not need to care when testing where the pages start and end.
 - write_pattern - writes a pattern of (0x00 0x01 .. 0xFF 0x00 ..) with the input length. uses unaligned writing. The advantage is the tester is not limited by shell input.
 - print 16 bytes/line (previous 8 bytes/line) to show large memory areas.

Tested on top of pull:
#26609 - LPC54114, 256bytes page, see command list below

 - erase a large area
flash erase 0x1FF00 1024
flash read 0x1FF00 1024

 - 3 pages: start at the end of first page, 1 full page, first part of last page
flash write_pattern 0x1FFF0 512
flash read 0x1FF00 1024

 - 1 single full page
flash write_pattern 0x20000 256
flash read 0x1FF00 1024

 - 1 single page first part written
flash write_pattern 0x20000 16
flash read 0x1FF00 512

 - 1 single page last part written
flash write_pattern 0x200F4 12
flash read 0x20000 256

 - 2 pages: firts page last part, last page first part
flash write_pattern 0x200F4 24
flash read 0x20000 512

agansari committed:
samples: flash_shell: unaligned and pattern write

Adds unaligned and pattern write functionality.
The reason this was added is chips with large pages, e.g. 256 can't be
normaly writen via shell, thus the need to write a pattern with any
length to a memory area.
The unaligned write command allows writing a data frament in any place
in memory, the data backup, protection, erasing and writing is
automaticaly handled. This makes it easier to test all flash controller
functionality in one command, also write any persistent data to any
flash address.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
samples: flash_shell: print 16 bytes of flash

On large memory areas printing 8 bytes/line is not enough.
Increased printing 16 bytes per line.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

d3zd3z commented on samples/drivers/flash_shell/prj.conf:
```diff
@@ -7,3 +7,4 @@ CONFIG_FLASH=y
 # If that's the case and you're interested in the flash layout, enable
 # it here.
 # CONFIG_FLASH_PAGE_LAYOUT=y
+CONFIG_HEAP_MEM_POOL_SIZE=16384
```
This file is missing a newline at the end.

d3zd3z commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -227,6 +248,140 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
+		before_data = k_malloc(page_size);
+		after_data = k_malloc(page_size);
+
+		if (!before_data || !after_data) {
```
If `before_data` allocation works, but `after_data` fails, this will leak the allocation of `before_data`.

d3zd3z commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -227,6 +248,140 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
+		}
+	}
+
+	ret = flash_write_protection_set(flash_device, true);
```
The previously allocated buffers aren't ever freed. They should be freed, no matter the return path. Doing them at the end, and having some of the error cases uses 'goto' to get down here is a common approach.

agansari replied:
@d3zd3z i've made the change to free buffers every time.

d3zd3z commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -227,6 +248,143 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
+	off_t  last_page = start_page + aligned_size - page_size;
+	bool single_page_write = (size_before + len < page_size);
+
+	static char *before_data;
```
Any reason these are static? Leftover from when they were buffers in RAM?

agansari replied:
You are right it's a leftover from when allocating buffers once.

d3zd3z commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -227,6 +248,143 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
+		return ret;
+	}
+
+	if (read_back) {
```
This does do the read back, even if something previously failed. Maybe the cleanup should be after this?

agansari replied:
Okay, will move the cleanup at the end of the function.

d3zd3z commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -336,7 +494,58 @@ static int cmd_write(const struct shell *shell, size_t argc, char **argv)
+	if (!buf) {
+		PR_ERROR(shell, "No heap memory for data pattern\n");
+		err = -ENOMEM;
+		goto free_buffers;
```
This error path only happens if buf is NULL, so there isn't much reason to go to a free_buffers label, exit would be fine.

agansari replied:
You're right, will remove label.

jakub-uC commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -227,6 +248,141 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
+}
+
+
 static int cmd_flash(const struct shell *shell, size_t argc, char **argv)
```
This command handler is obsolete. Would you mind removing it within this PR? 
Please also update command definition to:
`SHELL_CMD_REGISTER(flash, &sub_flash, "Flash realated commands.", NULL);`

Either way help message will be printed and we will save some code space.

agansari replied:
@jakub-uC i've added another commit removing the obsolete function.

agansari committed:
samples: flash_shell: remove obsolete command function

Removes 'flash' command function that displays the command's subset.
Command's subset will still be displayed in current code.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>
